cpp/velox/benchmarks/HashJoinBenchmark.cc

```diff
@@ -0,0 +1,307 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <arrow/filesystem/api.h>
+#include <arrow/result.h>
+#include <arrow/util/range.h>
+#include <benchmark/benchmark.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/file_reader.h>
+
+#include "compute/VeloxPlanConverter.h"
+#include "compute/protobuf_utils.h"
+#include "jni/exec_backend.h"
+#include "utils/exception.h"
+
+class WrapperIterator {
+ public:
+  explicit WrapperIterator(arrow::RecordBatchVector batches)
+      : batches_(std::move(batches)), iter_(batches_.begin()) {}
+
+  std::shared_ptr<arrow::RecordBatch> Next() {
+    return iter_ == batches_.cend() ? nullptr : *iter_++;
+  }
+
+ private:
+  arrow::RecordBatchVector batches_;
+  std::vector<std::shared_ptr<arrow::RecordBatch>>::const_iterator iter_;
+};
+
+std::shared_ptr<gluten::RecordBatchResultIterator> GetInputIterator(
+    const std::string& path) {
+  std::unique_ptr<parquet::arrow::FileReader> parquet_reader;
+  std::shared_ptr<arrow::RecordBatchReader> rb_reader;
+  parquet::ArrowReaderProperties properties = parquet::default_arrow_reader_properties();
+
+  GLUTEN_THROW_NOT_OK(parquet::arrow::FileReader::Make(
+      arrow::default_memory_pool(), parquet::ParquetFileReader::OpenFile(path),
+      properties, &parquet_reader));
+  GLUTEN_THROW_NOT_OK(parquet_reader->GetRecordBatchReader(
+      arrow::internal::Iota(parquet_reader->num_row_groups()), &rb_reader));
+  GLUTEN_ASSIGN_OR_THROW(auto collect_batches, rb_reader->ToRecordBatches());
+
+  return std::make_shared<gluten::RecordBatchResultIterator>(
+      std::make_shared<WrapperIterator>(std::move(collect_batches)));
+}
+
+auto BM_HashJoin = [](::benchmark::State& state, const std::string& l_input_file,
+                      const std::string& r_input_file) {
+  auto l_input_iter = GetInputIterator(l_input_file);
+  auto r_input_iter = GetInputIterator(r_input_file);
+
+  auto maybe_plan = SubstraitFromJSON("Plan", R"({
```

Comment:
Just a question, why not save the json file in a resource directory and read it back?

Reply:
For benchmarking purpose, the resource files can be very large. I can upload small files as an example.

Reply:
Actually I mean the json file of the Substrait plan, not the dataset. Do you mean the json file is also very large?

Reply:
Sorry for my misunderstanding. I think there's no need to put the json string in another file. If inline the json string mess the code, we can just extract it.

Reply:
I guess it would be more readable if extrating it from file. Velox is also using this way:
https://github.com/facebookincubator/velox/blob/main/velox/substrait/tests/JsonToProtoConverter.cpp
https://github.com/facebookincubator/velox/tree/main/velox/substrait/tests/data

Reply:
Thanks for the advice. Let's follow this practice.